packages/workflow/src/TelemetryHelpers.ts

```diff
@@ -525,3 +526,59 @@ export const userInInstanceRanOutOfFreeAiCredits = (runData: IRun): boolean => {
 
 	return false;
 };
+
+export type FromAICount = {
+	aiNodeCount: number;
+	aiToolCount: number;
+	fromAIOverrideCount: number;
+	fromAIExpressionCount: number;
+};
+
+export function makeAIMetrics(nodes: INode[], nodeTypes: INodeTypes): FromAICount | {} {
+	const resolvedNodes = nodes
+		.map((x) => [x, nodeTypes.getByNameAndVersion(x.type, x.typeVersion)] as const)
+		.filter((x) => !!x[1]?.description);
+
+	const aiNodeCount = resolvedNodes.reduce(
+		(acc, x) => acc + Number(x[1].description.codex?.categories?.includes('AI')),
+		0,
+	);
+
+	if (aiNodeCount === 0) return {};
+
+	let fromAIOverrideCount = 0;
+	let fromAIExpressionCount = 0;
+
+	const tools = resolvedNodes.filter((x) =>
+		x[1].description.codex?.subcategories?.AI?.includes('Tools'),
+	);
+
+	for (const [node, _] of tools) {
+		// FlatMap to support values in resourceLocators
+		const vals = Object.values(node.parameters).flatMap((x) => {
+			if (x && typeof x === 'object' && 'value' in x) x = x.value;
+			return typeof x === 'string' ? x : [];
+		});
+		// Note that we don't match the i in `fromAI` to support lower case i (though we miss fromai)
+		const overrides = vals.reduce(
+			(acc, x) => acc + Number(x.startsWith(`={{ ${FROM_AI_AUTO_GENERATED_MARKER} $fromA`)),
+			0,
+		);
+		fromAIOverrideCount += overrides;
+		// check for = to avoid scanning lengthy text fields
+		fromAIExpressionCount +=
+			vals.reduce((acc, x) => acc + Number(x[0] === '=' && x.includes('$fromA', 2)), 0) - overrides;
+	}
+
+	return {
+		aiNodeCount,
+		aiToolCount: tools.length,
+		fromAIOverrideCount,
+		fromAIExpressionCount,
+	};
+}
+
+export function isTool(nodeTypes: INodeTypes, node: INode): boolean {
+	const x = nodeTypes.getByNameAndVersion(node.type, node.typeVersion);
```

Comment:
Please use names instead of one letters as variable names, because it's clearer
```suggestion
	const nodeType = nodeTypes.getByNameAndVersion(node.type, node.typeVersion);
```